autoPyTorch/utils/implementations.py

```diff
@@ -61,3 +65,76 @@ def __call__(self, y: Union[np.ndarray, torch.Tensor]) -> np.ndarray:
     @staticmethod
     def get_properties() -> Dict[str, Any]:
         return {'supported_losses': ['BCEWithLogitsLoss']}
+
+
+class MinorityCoalescing(BaseEstimator, TransformerMixin):
+    """ Group together categories which occurence is less than a specified
+    minimum fraction. Coalesced categories get index of one.
+    """
+
+    def __init__(self, minimum_fraction: Optional[float] = None):
+        self.minimum_fraction = minimum_fraction
+
+    def check_X(self, X: np.array) -> None:
+        X_data = X.data if sparse.issparse(X) else X
+        if np.nanmin(X_data) <= -2:
+            raise ValueError("X needs to contain only integers greater than -2.")
+
+    def fit(self, X: np.array, y: Optional[np.ndarray] = None) -> 'MinorityCoalescing':
+        self.check_X(X)
+
+        if self.minimum_fraction is None:
+            return self
+
+        # Remember which values should not be coalesced
+        do_not_coalesce: List[Set[int]] = list()
+        for column in range(X.shape[1]):
+            do_not_coalesce.append(set())
+
+            if sparse.issparse(X):
+                indptr_start = X.indptr[column]
+                indptr_end = X.indptr[column + 1]
+                unique, counts = np.unique(
+                    X.data[indptr_start:indptr_end], return_counts=True)
+                colsize = indptr_end - indptr_start
+            else:
+                unique, counts = np.unique(X[:, column], return_counts=True)
+                colsize = X.shape[0]
+
+            for unique_value, count in zip(unique, counts):
+                fraction = float(count) / colsize
+                if fraction >= self.minimum_fraction:
+                    do_not_coalesce[-1].add(unique_value)
```

Comment:
I found some issues and refactored the codes.
**Issues**
1. colsize changes depending on if the feature matrix is sparse or not (`indptr_end - indptr_start` is not correct)
2. Unneeded usage of for loop (rewrite by numpy to eliminate for loop, which I did)

This function can also be written nicely.
```suggestion
        n_instances, n_features = X.shape
        do_not_coalesce = [set() for _ in range(n_features)]
        is_sparse = sparse.issparse(X)

        for col, no_coal in enumerate(do_not_coalesce):
            if is_sparse:
                indptr_start = X.indptr[column]
                indptr_end = X.indptr[column + 1]
                col_data = X.data[indptr_start:indptr_end]
            else:
                col_data = X[:, col]

            unique_vals, counts = np.unique(col_data, return_counts=True)
            frac = counts / n_instances
            no_coal.update(unique_vals[frac >= min_frac])
```

Reply:
Yes, but we do not want to merge fit and transform as we are using the scikit-learn API.  You can read more about it [here](https://towardsdatascience.com/fit-vs-transform-in-scikit-libraries-for-machine-learning-3c70e6300ded)

Reply:
Ok, then I will respond to the corresponding lines separately.